### PR TITLE
Unattended

### PR DESCRIPTION
**Pull request by jlagermann**
**By submitting this pull request, I confirm the following:** 
*please fill any appropriate checkboxes, e.g: [X]*

- [x] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md), as well as this entire template.
- [x] I have made only one major change in my proposed changes.
- [x] I have commented my proposed changes within the code.
- [x] I have tested my proposed changes, and have included unit tests where possible.
- [x] I am willing to help maintain this change if there are issues with it later.
- [x] I give this submission freely and claim no ownership.
- [x] It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
- [x] I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))

Please make sure you [Sign Off](https://github.com/pi-hole/pi-hole/wiki/How-to-signoff-your-commits.) all commits. Pi-hole enforces the [DCO](https://github.com/pi-hole/pi-hole/wiki/Contributing-to-the-project).

---
**What does this PR aim to accomplish?:**
Using the --unattended flag to install requires the use of setupVars.conf.  If this configuration file is missing required field or not formatted correctly, the installation will fail, with or without displaying errors.

This PR request adds validation for setupVars.conf, adds an example setupVars.conf, and creates a functions to create or update setupVars.conf or any other configuration file.

The function replaceAddCreate will replace existing variables while maintaining any comments and indentation or add variables if they do not exist.  It will also create the file with the correct permissions and format.

The function replaceAddCreateRename will do the same as replaceAddCreate but also allows the ability to rename the variable in the target file.  For example, DNS_CACHE_SIZE is named cache-size in 01-pihole.conf.

This PR also made other minor changes related to the use of setupVars.conf.  

To learn the random password that is created during an unattended installation, displayFinalMessage is enabled for --unattended, with the exception of whiptail message.  Pipe the output to a log file, for example "~/pihole-install.sh --unattended >> /var/log/unattended-pihole-install.log"

If useUpdateVars=true, it will first make sure setupVars.conf exists and has the correct permissions, then run accountForRefactor, then validate_setupVars.

It checks to make sure INSTALL_WEB_SERVER is enabled if INSTALL_WEB_INTERFACE is enabled is (part of the validate_setupVars).

If the undocumented flag "--disable-install-webserver" is used, both INSTALL_WEB_INTERFACE and INSTALL_WEB_SERVER will be set to false.

Initial setting of INSTALL_WEB_SERVER=true was moved up with other variables.

added support for PR #3170.  Any variable can be added by creating a default in basic-install.sh, adding the variable in all caps to pihole-setupVars.conf, and setting the variable in the appropriate file with "replaceAddCreate" or "replaceAddCreateRename", depending on the name of the variable in the target file.  If the variable name has to be changed, use "replaceAddCreateRename".

Also fixed a typo in FTLinstall (changed "grep Location" to "grep location")

**How does this PR accomplish the above?:**
- added pihole-setupVars.conf example
- added validate_setupVars
- added replaceAddCreate
- added replaceAddCreateRename
- updated finalExports to use replaceAddCreate
- validate_setupVars runs immediately after accountForRefactor

**What documentation changes (if any) are needed to support this PR?:**
- DNS Cache documentation should be updated with instruction to add changes to setupVars.conf
- Installation instruction should be updated for unattended installs.
- 01-pihole.conf should be updated with a note about making changes.

---
* You must follow the template instructions. Failure to do so will result in your pull request being closed.
* Please respect that Pi-hole is developed by volunteers, who can only reply in their spare time.

### CODE REVIEW

**Comment by dschaper** [comment]
I'll be honest. The chances of me approving this is very small. Too many complex changes made in one PR. Each PR needs to do one specific task with an explanation of what along with, more importantly, why.

File: automated install/basic-install.sh [hunk 1]
```diff
@@ -2229,7 +2363,7 @@ FTLinstall() {
     printf "  %b %s..." "${INFO}" "${str}"
 
     # Find the latest version tag for FTL
-    latesttag=$(curl -sI https://github.com/pi-hole/FTL/releases/latest | grep "Location" | awk -F '/' '{print $NF}')
+    latesttag=$(curl -sI https://github.com/pi-hole/FTL/releases/latest | grep "location" | awk -F '/' '{print $NF}')
```
Comment (dschaper) [inline automated install/basic-install.sh]: This is different that what is in master. Master already has `grep -i` to account for capitalization.

https://github.com/pi-hole/pi-hole/blob/master/automated%20install/basic-install.sh#L2387

File: automated install/basic-install.sh [hunk 2]
```diff
@@ -1825,7 +1832,51 @@ configureFirewall() {
     printf "  %b Skipping firewall configuration\\n" "${INFO}"
 }
 
-#
+replaceAddCreateRename() {
+    # origVariable="$1"
+    # newVariable="$2"
+    # theFile="$3"
+    # If the file does not exist and is not larger than 0 bytes, create it with 644 permisions or add 1 byte. sed cannot work on empty files
+    if [[ ! -s "$3" ]] ; then
```
Comment (dschaper) [inline automated install/basic-install.sh]: Use braces in variables please.

File: automated install/basic-install.sh [hunk 3]
```diff
@@ -1825,7 +1832,51 @@ configureFirewall() {
...
+replaceAddCreate() {
+    # theVariable="$1"
+    # theValue="${!1}"
+    # theFile="$2"
+    # If the file does not exist and is not larger than 0 bytes, create it with 644 permisions or add 1 byte. sed cannot work on empty files
+    if [[ ! -s "$2" ]] ; then
+       (umask 022 ; echo >> "$2")
+    fi
+    # Replace the value of a variable or add it to a target file if it is not present, preserving indentation and comments.
+    sed -i '/^[#]*[ \t]*'"$1"'=/ {h;s/=[^#]*\(.*\)/='"${!1}"'\1/;s/'"${!1}"'#/'"${!1}"' #/};${x;/^$/{s//'"$1"'='"${!1}"'/;H};x}' $2
```
Comment (dschaper) [inline automated install/basic-install.sh]: Break this in to multiple steps. No need to oneline it.

File: automated install/basic-install.sh [hunk 4]
```diff
@@ -1907,16 +1948,112 @@ accountForRefactor() {
...
+    for i in "${requiredVars[@]}"; do
+        if [[ -z "${!i}" ]]; then
+            echo "$i is not set"
+            exit 1
+        else
+            echo "$i is ${!i}"
+        fi
+    done
+    for i in "${optionalVarsTF[@]}"; do
+        if [[ "${!i}" == "true" ]] || [[ "${!i}" == "false" ]]; then
```
Comment (dschaper) [inline automated install/basic-install.sh]: Test for a positive condition, not the negation of a variable.

Comment (dschaper) [reply automated install/basic-install.sh]: NVM, I see what you are doing.

File: automated install/basic-install.sh (hunk 3 above)
Comment (jlagermann) [reply automated install/basic-install.sh]: done

File: automated install/basic-install.sh (hunk 2 above)
Comment (jlagermann) [reply automated install/basic-install.sh]: done

File: automated install/basic-install.sh [hunk 5]
```diff
@@ -1833,48 +1833,71 @@ configureFirewall() {
 }
 
 replaceAddCreateRename() {
-    # origVariable="$1"
-    # newVariable="$2"
-    # theFile="$3"
+    # origVariable="${1}"
+    # newVariable="${2}"
+    # theFile="${3}"
```
Comment (dschaper) [inline automated install/basic-install.sh]:
```suggestion
local theFile
theFile="${3}"
```

File: automated install/basic-install.sh [hunk 6]
```diff
@@ -1833,48 +1833,71 @@ configureFirewall() {
...
     # If the file does not exist and is not larger than 0 bytes, create it with 644 permisions or add 1 byte. sed cannot work on empty files
-    if [[ ! -s "$3" ]] ; then
-       (umask 022 ; echo >> "$3")
+    if [[ ! -s "${3}" ]] ; then
```
Comment (dschaper) [inline automated install/basic-install.sh]:
```suggestion
    if [[ ! -s "theFile" ]] ; then
```

File: automated install/basic-install.sh (hunk 1 above)
Comment (jlagermann) [reply automated install/basic-install.sh]: okay, I'm trying to figure out how to resolve the conflict.  I did 'git fetch upstream' and my copy of development still does not have grep -i.

File: automated install/basic-install.sh (hunk 5 above)
Comment (jlagermann) [reply automated install/basic-install.sh]: origVariable, newVariable, and theFile are just there for illustration purposes, there not used.  That's why I commented them out.

Comment (jlagermann) [reply automated install/basic-install.sh]: I'm good with using them if you want.

Comment (dschaper) [reply automated install/basic-install.sh]: Please use them throughout. There shouldn't be any positional variables or iterators that are named with the numeric position or a single letter like `${i}`.

Comment (jlagermann) [reply automated install/basic-install.sh]: I only use ${i} in for loops.  I'll change them to ${var}.

**Comment by PromoFaux** [comment]
@jlagermann I'm a little worried about how out of date your `development` branch is with the main repo. E.g there are commits from as far back as january missing from your end (i'll do a pull request of `pi-hole:development` into `jlagermann:unattended` so you can see.

Greatful for the contribution, it is always good to have another set of eyes on `basic-install.sh`, but there are a few things missing which could cause issues if we accidentally lose them with an incorrect merge.

**Comment by jlagermann** [comment]
I thought it was odd my branch is as far off as it is, I just forked the whole repo 10 days ago.
What's the best course of action from here?  I'm okay with:
- deleting this branch, 
- merging pi-hole:development with jlagermann:development, 
- recreating the branch,
- coping my changes back into setupVars.conf
- testing
- PR

If there is a shorter path, I'm open to it.

**Comment by PromoFaux** [comment]
Personally, what I would do when contributing to a project:
 - Make a fresh fork of the project
 - Clone my fork locally
 - Create a new branch based on the development branch (`git checkout development && git checkout -b new/mynewfeature`)
 - do the work, commit it, push back to my fork (`git push origin new/mynewfeature`)
 - Submit a pull request against the original repo on github
 - once PR is accepted, delete my own fork - I don't need it until I have something new to contribute

I just wonder if the extra steps you have with mirroring to gitlab etc are causing issues.

**Comment by jlagermann** [comment]
I'm not actually mirroring to gitlab for this project.  About 90% of my
work is in a gitlab that I manage.  At Corelight, our developers work in
gitlab and then mirror their published work to github for the public.

In this case, I did a fresh fork for this PR and a documentation PR.  It
just took me about 2 weeks to get through all the testing I did.  I really
don't understand how it's missing so many commits.  If you look at PR
#3178, that is on the same fork, just a different branch on my side.

Before we close this PR, later this week I'll do a fresh fork and copy my
work over to it.  Then do a new PR.



Thank you,

James Lagermann
636-575-7249

**Comment by PromoFaux** [comment]
In the case of the other PR, you're only touching one file, which hasn't been touched by any of the later commits, so that is fine. 

In the case of this one, the file you've made changes to has also been changed since you originally forked. 

You may even get away with just rebasing this branch on master... If you have not already added an origin for the original repo:

```
git remote add upstream https://github.com/pi-hole/pi-hole.git
git fetch upstream
git rebase upstream/development
```

File: automated install/basic-install.sh [hunk 7]
```diff
@@ -109,14 +111,13 @@ c=$(( c < 70 ? 70 : c ))
...
 runUnattended=false
-INSTALL_WEB_SERVER=true
 # Check arguments for the undocumented flags
 for var in "$@"; do
     case "$var" in
         "--reconfigure" ) reconfigure=true;;
         "--i_do_not_follow_recommendations" ) skipSpaceCheck=true;;
         "--unattended" ) runUnattended=true;;
-        "--disable-install-webserver" ) INSTALL_WEB_SERVER=false;;
+        "--disable-install-webserver" ) INSTALL_WEB_SERVER=false INSTALL_WEB_INTERFACE=false;;
```
Comment (MichaIng) [inline automated install/basic-install.sh]: When the webserver install is skipped, this does not mean that the user does not want the web interface. It is possible that Apache or Nginx is used as webserver already and the user wants to continue is it. So if it is wanted to skip the dialog, a separate cmd option is required IMO.

**Comment by MichaIng** [comment]
Hey guys, I like the idea to allow a completely unattended install, either via command line options or when having a `setupVars.conf` in place. But I see the vast other changes/additions done here. What do you think about the following:

- Add command line options for all questions that are otherwise asked for during install. `--disable-install-webserver` can be taken as an example, to do this in a consistent way.
  - The options must be given the way that they imply the non-default choice, if it is yes/no.
  - If an option is present, the related whiptail question is skipped.
  - If the `--unattended` option is given, all whiptail questions are skipped now, at least if the `setupVars.conf` is present.
- Adjust the `--unattended` option to allow adding an optional config file path: `--unattended=/path/to/my.conf`
  - If no path to config file is given with the `--unattended` option, it expects and uses the default location to keep existing behaviour.
  - If neither given, nor default file path exists, exit with error message.
- If any command line option is given, the values from config file are overridden.
- If the `--unattended[=/path/to/my.conf]` option is used, but neither the config file, nor a command line option defines a certain choice, the default value is used. Probably in some cases, default values need to be added to code.

A documentation about config file options/syntax would be nice, however for now it allows to use the file from one install, to do future or further installs automated with it.

Code structure would need to be adjusted like this:
0. Declare all choice/settings variables and assign default values
1. Source `setupVars.conf` if `--unattended` is given + given or default file path exists
2. Parse command line options and, in case, overwrite variables
3. Skip all whiptail options if `--unattended` is given, use the given values, which are either from command line options (1. priority) or config file (2. priority) or default value

**Comment by jlagermann** [comment]
I'm not sure I see the value in making all the variables available via the
command line and not use the setupVars files.  At a minimum you would have
4 variables you would have to pass supply, in addition to the
"--unattended" option.

I completely agree with automated installations.  I use an Ansible playbook
to install and setup Pi-hole.  I keep it in my gitlab repository.  Anytime
I need to make a change, I change it in the repository.  Once it's
committed, a pipeline kicks off to update Pi-hole.

As far as the documentation, I think I have that covered with the example
"pihole-setupVars.conf.example".  If you wanted to supply all the variables
via the command line I've documented so far, you would have 20 of them.  I
can provide my basic Ansible playbook but I was not going to upload it with
this PR.

If you did use the command line, the setupVars.conf file would override any
duplicate variables.  The command line options would be evaluated when the
script runs.  The variables provided via setupVars.conf are not imported
until after all of the command line flags.

Most of the changes in my PR have to do with validating the user input.  A
lot of things can go wrong if the variables are not provided correctly.
For example, if they say "yes" instead of "true", or if the IP address does
not include a subnet mask in the form of /xx.

As simple as it sounds, the path to setupVars.conf is used in multiple
places in pi-hole.  I know the file is also the PHP ini settings file, I'm
not sure if it is used in other places.  You could provide a separate file
just for the initial unattended setup, but it would still have to create
setupVars.conf in the proper place and write all the variables to it.  It's
also used for updates (basically reinstalls and keeps the existing
settings).


Thank you,

James Lagermann
636-575-7249

**Comment by MichaIng** [comment]
@jlagermann 
> I'm not sure I see the value in making all the variables available via the
command line and not use the setupVars files.  At a minimum you would have
4 variables you would have to pass supply, in addition to the
"--unattended" option.

See them as an optional addition to force/override a specific value from config file in case you use `--unattended` already. But the most obvious use case is making the regular install partly or completely unattended without a config file. Compare it with other commands like certbot, which is natively interactive, but allows to answer some or all choices via command line options.

> As far as the documentation, I think I have that covered with the example
"pihole-setupVars.conf.example".  If you wanted to supply all the variables
via the command line I've documented so far, you would have 20 of them.  I
can provide my basic Ansible playbook but I was not going to upload it with
this PR.

Yes, this works already, but it requires the `setupVars.conf` at a specific place and most values must be present and valid since not all variables are init with reasonable defaults. It would be more handy if one could only pass/force a few options while falling back to working defaults (`--unattended`) or interactive whiptail choices for the other options.

> If you did use the command line, the setupVars.conf file would override any
duplicate variables.  The command line options would be evaluated when the
script runs.  The variables provided via setupVars.conf are not imported
until after all of the command line flags.

Yes this would need to be changed in order, as of my post above.

> Most of the changes in my PR have to do with validating the user input.  A
lot of things can go wrong if the variables are not provided correctly.
For example, if they say "yes" instead of "true", or if the IP address does
not include a subnet mask in the form of /xx.

Yes that is true and indeed validating correct values is a great addition. In case of "yes!=true!=1" etc it is of course up to the user to provide correct entries + syntax. The way checks are done, an invalid value should lead to default value being used. In case a string input is required, a check would be best, although the interactive installer also does not validate e.g. custom upstream DNS IP, does it?

> As simple as it sounds, the path to setupVars.conf is used in multiple
places in pi-hole.  I know the file is also the PHP ini settings file, I'm
not sure if it is used in other places.  You could provide a separate file
just for the initial unattended setup, but it would still have to create
setupVars.conf in the proper place and write all the variables to it.  It's
also used for updates (basically reinstalls and keeps the existing
settings).

Yes, basically the custom config file must be parsed only once at the top of script, then it must not be used anymore. As it is already now, at later stage, all final choices/values are stored to the regular  `setupVars.conf` location for pihole -up/-r.

**Comment by jlagermann** [comment]
I'm recreating this PR with multiple pull requests (all dependent on each other).  The first PR is #3589 .  More will follow.

**Pull request closed without merging**